src/services/ftp/backend.rs

```diff
@@ -0,0 +1,512 @@
+            port,
+            root,
+            credential,
+            tls,
+        })
+    }
+}
+
+/// Backend is used to serve `Accessor` support for ftp.
+#[derive(Clone)]
+pub struct Backend {
+    endpoint: String,
+    port: String,
+    root: String,
+    credential: (String, String),
+    tls: bool,
+}
+
+impl Debug for Backend {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Backend")
+            .field("endpoint", &self.endpoint)
+            .field("root", &self.root)
+            .field("credential", &self.credential)
+            //.field("stream", &self.stream)
+            .finish()
+    }
+}
+
+impl Backend {
+    pub(crate) fn from_iter(it: impl Iterator<Item = (String, String)>) -> Result<Self> {
+        let mut builder = Builder::default();
+
+        for (k, v) in it {
+            let v = v.as_str();
+            match k.as_ref() {
+                "root" => builder.root(v),
+                _ => continue,
+            };
+        }
+        builder.build()
+    }
+    /*
+    pub(crate) fn get_rel_path(&self, path: &str) -> String {
+        match path.strip_prefix(&self.root) {
+            Some(v) => v.to_string(),
+            None => unreachable!(
+                "invalid path{} that does not start with backend root {}",
+                &path, &self.root
+            ),
+        }
+    }
+
+    pub(crate) fn get_abs_path(&self, path: &str) -> String {
+        if path == "/" {
+            return self.root.to_string();
+        }
+        format!("{}{}", self.root, path)
+    }
+    */
+}
+
+#[async_trait]
+impl Accessor for Backend {
+    async fn create(&self, args: &OpCreate) -> Result<()> {
+        let path = args.path();
+
+        if args.mode() == ObjectMode::FILE {
+            self.ftp_put(path).await?;
+
+            return Ok(());
+        }
+
+        if args.mode() == ObjectMode::DIR {
+            self.ftp_mkdir(path).await?;
+
+            return Ok(());
+        }
+
+        unreachable!()
+    }
+
+    async fn read(&self, args: &OpRead) -> Result<BytesReader> {
+        let path = args.path();
+
+        let reader = self.ftp_get(path).await?;
+
+        let mut reader = Compat::new(reader);
+
+        if let Some(offset) = args.offset() {
+            reader
+                .seek(SeekFrom::Start(offset))
+                .await
+                .map_err(|e| parse_io_error(e, "read", args.path()))?;
+        }
+
+        let r: BytesReader = match args.size() {
+            Some(size) => Box::new(reader.take(size)),
+            None => Box::new(reader),
+        };
+
+        Ok(r)
+    }
+
+    async fn write(&self, args: &OpWrite, r: BytesReader) -> Result<u64> {
+        let path = args.path();
+
+        let n = self.ftp_append(path, unshared_reader(r)).await?;
+
+        Ok(n)
+    }
+
+    async fn stat(&self, args: &OpStat) -> Result<ObjectMetadata> {
+        let path = args.path();
+
+        if path == self.root {
```

Comment:
You may misunderstood how `root` in OpenDAL works. :(

Users can use `root` to set the working directory of OpenDAL performs in.

For example, if you set `root` to be `/path/to/dir`, then if you create a file named `example.txt` (via OpenDAL, of course), its real, absolute path will be `/path/to/dir/example.txt`.